lib/descriptor/common.js

```diff
@@ -0,0 +1,259 @@
+/*!
+ * common.js - common functions for descriptor
+ * Copyright (c) 2023, the bcoin developers (MIT License).
+ * https://github.com/bcoin-org/bcoin
+ */
+
+'use strict';
+
+const assert = require('bsert');
+const {types} = require('../script/common');
+const BN = require('bcrypto/lib/bn.js');
+const common = exports;
+
+/**
+ * parse script context for descriptor
+ * @const {Object}
+ */
+
+common.scriptContext = {
+  TOP: 'TOP',
+  P2SH: 'P2SH',
+  P2WPKH: 'P2WPKH',
+  P2WSH: 'P2WSH',
+  P2TR: 'P2TR'
+};
+
+/**
+ * Derive type for a key object in a descriptor
+ * @const {Object}
+ */
+
+common.deriveType = {
+  NO: 'NO',
+  UNHARDENED: 'UNHARDENED',
+  HARDENED: 'HARDENED'
+};
+
+/**
+ * String type for key object in a descriptor
+ * @const {Object}
+ */
+
+common.stringType = {
+  PUBLIC: 'PUBLIC',
+  PRIVATE: 'PRIVATE',
+  NORMALIZED: 'NORMALIZED'
+};
+
+/**
+ * Output types for descriptor
+ * @const {Object}
+ */
+
+common.outputTypes = types;
+
+/**
+ * Valid input characters for input in a descriptor
+ * and used for calculating checksum
+ * @const {String}
+ */
+
+common.INPUT_CHARSET = '0123456789()[],\'/*abcdefgh@:$%{}IJKLMNOPQRSTUVWXYZ' +
+                       '&+-.;<=>?!^_|~ijklmnopqrstuvwxyzABCDEFGH`#"\\ ';
+
+ /**
+ * Checksum character set for calculating checksum
+ */
+
+common.CHECKSUM_CHARSET = 'qpzry9x8gf2tvdw0s3jn54khce6mua7l';
+
+/**
+ * Test whether this descriptor is of scriptType
+ * ('pk', 'pkh', 'wpkh' etc.) or not
+ * @param {String} scriptType type of script
+ * @param {String} desc descriptor string
+ * @returns {Boolean}
+ */
+
+common.descType = function descType(scriptType, desc) {
+  if (
+    desc.length >= scriptType.length + 2 &&
+    desc[scriptType.length] === '(' &&
+    desc[desc.length - 1] === ')' &&
+    scriptType === desc.substring(0, scriptType.length)
+  ) {
+    return true;
+  }
+  return false;
```

Comment:
Simplify and add some comments explaining the logic

```suggestion
  return desc.length >= scriptType.length + 2 &&
    desc[scriptType.length] === '(' &&
    desc[desc.length - 1] === ')' &&
    scriptType === desc.substring(0, scriptType.length);
```